File: cpp/DecodeWays.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

class Solution {
public:
    int numDecodings(string s) {
        if (s[0] == '0') {
            return 0;
        }
        
        int n = s.size();
        
        vector<int> dp(n + 1);
        dp[0] = 1;
        dp[1] = 1;
        
        for (int i = 2; i <= n; i++) {
            int digits = stoi(s.substr(i - 1, 1));
            if (digits >= 1 && digits <= 9) {
                dp[i] += dp[i - 1];
            }
            int tens = stoi(s.substr(i - 2, 2));
            // the code must be less than 26 and greater than 0, since there are only 26 letters
            if (tens >= 10 && tens <= 26) {
                dp[i] += dp[i - 2];
            }
        }
        
        return dp[n];
    }
};

File: python/DecodeWays.py
class Solution:
    def numDecodings(self, s: str) -> int:
        # caching
        cache = { len(s) : 1 }
        def dfs(i):
            if i in cache:
                return cache[i]
            if s[i] == "0":
                return 0

            res = dfs(i + 1)
            # if char is one, then char after it can be any, but if its 2, it can only be 1 thru 6, since there are 26 letters in the alphabet
            if (i + 1 < len(s) and (s[i] == "1" or (s[i] == "2" and s[i + 1] in "0123456") )):
                res += dfs(i + 2)
            
            cache[i] = res
            return res

        return dfs(0)